tga-runner/src/main/kotlin/org/plan/research/tga/runner/coverage/jacoco/JacocoCoverageProvider.kt

```diff
@@ -60,13 +61,59 @@ class JacocoCoverageProvider(
     }
 
     override fun computeCoverage(benchmark: Benchmark, testSuite: TestSuite): ClassCoverageInfo {
+        log.debug("Computing coverage: compiledDir='{}'", compiledDir)
+
+        // set of sources containing both test cases and the test suite
         val allTests = testSuite.tests.associateWith { testSuite.testSrcPath.resolve(it.asmString + ".java") }
+        // set of sources containing the single test suite
+        val testSuiteOnly = if (testSuite.testSuiteQualifiedName.isNotEmpty()) {
+                testSuite.testSuiteQualifiedName.let { testSuiteName ->
+                    listOf(testSuiteName).associateWith { testSuite.testSrcPath.resolve(it.asmString + ".java") }
+                }
+            } else {
+                null
+            }
+        // set of sources containing only test cases without the test suite
+        val testCasesOnly = if (testSuite.testCasesOnly.isNotEmpty()) {
+                testSuite.testCasesOnly.associateWith { testSuite.testSrcPath.resolve(it.asmString + ".java") }
+            } else {
+                null
+            }
+
+        /**
+         * Provide up to 3 sets of source files suitable for compilation and
+         * try to collect coverage for at least one of them since all 3 represent
+         * the same coverage set:
+         * 1. All the compilable test cases and the test suite (test suite contains only compilable test cases).
+         * 2. All the compilable test case alone.
+         * 3. The test suite alone.
+         */
+        val compilationAttempts = mutableListOf(allTests)
+        testSuiteOnly?.also { compilationAttempts.add(it) }
+        testCasesOnly?.also { compilationAttempts.add(it) }
+
         val classPath = benchmark.classPath + testSuite.dependencies.flatMap { dependencyManager.findDependency(it) }
         val compiler = SystemJavaCompiler(classPath)
-        try {
-            compiler.compile(allTests.values.toList(), compiledDir)
-        } catch (e: Throwable) {
-            log.error(e)
+
+        var selectedTestSet: Map<String, Path> = emptyMap()
+
+        for ((index, attempt) in compilationAttempts.withIndex()) {
+            log.debug(
+                "Attempting to compile {}-th set of tests: [\n{}\n\t]",
+                index,
+                attempt.keys.joinToString(separator = "\n") { "\t\t$it," },
+            )
+
+            try {
+                val result = compiler.compile(attempt.values.toList(), compiledDir)
+                selectedTestSet = attempt
+                log.debug("Compilation succeeded with result: {}", result)
+                break
+            }
+            catch (e: Throwable) {
+                log.error(e)
+                // TODO: should compiledDir be cleaned up before next attempt?
```

Comment:
Some of the files may appear there (the ones that were successfully compiled before failure), so in the perfect world we can delete them. However, I don't think that the results will be somehow affected if we leave them in place.